p2p/src/channels/snark/p2p_channels_snark_reducer.rs

```diff
@@ -1,63 +1,119 @@
-use super::{
-    P2pChannelsSnarkAction, P2pChannelsSnarkActionWithMetaRef, P2pChannelsSnarkState,
-    SnarkPropagationState,
+use openmina_core::{bug_condition, Substate};
+use redux::ActionWithMeta;
+
+use crate::{
+    channels::snark_effectful::P2pChannelsSnarkEffectfulAction, P2pNetworkPubsubAction, P2pState,
 };
 
+use super::{P2pChannelsSnarkAction, P2pChannelsSnarkState, SnarkPropagationState};
+use mina_p2p_messages::{gossip::GossipNetMessageV2, v2};
+
 impl P2pChannelsSnarkState {
-    pub fn reducer(&mut self, action: P2pChannelsSnarkActionWithMetaRef<'_>) {
+    pub fn reducer<Action, State>(
+        mut state_context: Substate<Action, State, P2pState>,
+        action: ActionWithMeta<&P2pChannelsSnarkAction>,
+    ) -> Result<(), String>
+    where
+        State: crate::P2pStateTrait,
+        Action: crate::P2pActionTrait<State>,
+    {
         let (action, meta) = action.split();
+        let p2p_state = state_context.get_substate_mut()?;
+
+        let state = action
+            .peer_id()
+            .and_then(|peer_id| p2p_state.get_ready_peer_mut(peer_id))
+            .map(|peer_state| &mut peer_state.channels.snark)
+            .ok_or_else(|| format!("Invalid state for: {action:?}"));
+
         match action {
-            P2pChannelsSnarkAction::Init { .. } => {
-                *self = Self::Init { time: meta.time() };
+            P2pChannelsSnarkAction::Init { peer_id } => {
+                let state = state?;
```

Comment:
Are these bug-conditions? will they be correctly processed as such?

Reply:
These are bug conditions but I didn't add them in places like this, I will added them just in case